### DISCUSSION TITLE
Node REPL と ファイル実行で `this` は変更されますか？

### DISCUSSION THREAD

**Discussion by fdsa789** [asciidwango/js-primer#1483] [質問]
Node REPL と ファイルを実行したときに、`this` の戻り値が違うのが気になっています。

## 該当ページ
アロー関数の実行について
https://jsprimer.net/basic/function-this/#arrow-function-this

## 質問
以下は `node ファイル名.js` で実行した結果です。

```js
'use strict';

function funcThis() {
  return this;
}

console.log(funcThis()); // undefined

const arrowThis = () => {
  return this;
};

console.log(arrowThis()); // {}
console.log(arrowThis() === this); // true
console.log(arrowThis() === globalThis); // false
```

しかし、Node REPL で実行すると以下の結果になり、`arrowThis()` の戻り値は `global`  になっています。
ファイル名で実行したとき (node ファイル名.js) の戻り値は `{}` に対して、Node REPL では `global` が返されるのですが、どのような違いがあるのでしょうか。

また Node REPL の実行だと `arrowThis() === this` と `arrowThis() === globalThis` がどちらとも `true` ですが、ファイル名で実行すると `arrowThis() === globalThis` が `false` になっています。

![screenshot](https://user-images.githubusercontent.com/107389031/194691626-1675a25b-e77d-4c9e-949d-004847c94178.png)

**Comment by azu** [comment]
📝 調べ方としては `console.trace` などでスタックトレースをみると、どこから呼ばれているかわかるので、これを元にNodeのソースコードを追って見つけることができそう。 `node --inspect`でdebuggerを起動すると https://github.com/nodejs/node/blob/78d280a768219a6af484b7ecbf3e836461b57477/lib/internal/modules/cjs/loader.js#L1156-L1157 専用のモードで起動しちゃうのでちょっとスタックが変わってしまっていた。

<img width="748" alt="image" src="https://user-images.githubusercontent.com/19714/194891147-82b1554b-dfad-4415-a502-c17e685b5a68.png">

----

まだちゃんとは調べていませんが、Node.jsの`.js`実行はデフォルトではCommonJSとして実行されます。(`.mjs`とかにするとECMAScript Modulesになって`this`が変わるのが確認できる。mjsはECMAScriptの仕様通りに評価しないといけない。CommonJSでは globalの `this` をどうするかという仕様がないはず?)
このときにCommonJSとしての`exports`や`require`などの関数を参照できるようにファイルの内容を関数でラップしてから実行してるようです。

また、CommonJSファイルの実行は[Reflect.apply()](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Reflect/apply)(`Function.apply`と大体同じ)で、`thisValue` として `exports` を指定しているようです。
([Modules/1.1 - CommonJS Spec Wiki](https://wiki.commonjs.org/wiki/Modules/1.1)でこれが仕様として決まっているのかはよくわかってないです)

https://github.com/nodejs/node/blob/cebbc57ed26ce1a572e6fe416b700ac8f6258869/lib/internal/modules/cjs/loader.js#L253-L256
https://github.com/nodejs/node/blob/main/lib/internal/modules/cjs/loader.js#L1159-L1160

これは、次のようなファイルを `node test.js` とかで実行してみると `this` が `exports` を示していることがわかると思います。
これが `console.log(arrowThis() === globalThis); // false` の正体だと思います。(CommonJSはECMAScriptの仕様ではないので、CJSにおける `this` の正解は存在しないという理解)

```js
"use strict";
exports.test = 1;
function funcThis() {
  return this;
}

console.log(funcThis()); // undefined

const arrowThis = () => {
  return this;
};

console.log(arrowThis()); // { test: 1 }
console.log(arrowThis() === this); // true
console.log(arrowThis() === globalThis); // false
```

同じように REPLでも`console.trace` などで スタックトレースから辿ると、次のようなスタックトレースが出てくるので、[VM](https://nodejs.org/api/vm.html)を使って実行しているようです。

<img width="607" alt="image" src="https://user-images.githubusercontent.com/19714/194891007-0347543c-ef36-473f-b24d-13693d076897.png">


```
    at REPL15:1:9
    at Script.runInThisContext (node:vm:129:12)
    at REPLServer.defaultEval (node:repl:566:29)
```

`Script.runInThisContext` で実行するということは、vmの中でV8のisolationを作ってそこでeval的な感じで実行してるということなんだと思います。(前回実行した結果はcontextとして保持しておいて、次の行を実行するときにcontextを渡しながら新しい環境で実行を繰り返す)

https://github.com/nodejs/node/blob/78d280a768219a6af484b7ecbf3e836461b57477/lib/vm.js#L124-L130
→
https://github.com/nodejs/node/blob/78d280a768219a6af484b7ecbf3e836461b57477/src/node_contextify.cc#L736
→
https://github.com/nodejs/node/blob/78d280a768219a6af484b7ecbf3e836461b57477/src/node_contextify.cc#L915-L966
→
https://github.com/nodejs/node/blob/78d280a768219a6af484b7ecbf3e836461b57477/src/node_contextify.cc#L389-L390

V8をコードとEnvを渡して呼び出していて、おそらくEnvにthisの値となるものが指定されてると思うので、この辺を追うとわかるような気がします。

https://github.com/nodejs/node/blob/78d280a768219a6af484b7ecbf3e836461b57477/src/node_contextify.cc#L753

https://github.com/nodejs/node/blob/78d280a768219a6af484b7ecbf3e836461b57477/src/node_contextify.cc#L736

- https://github.com/nodejs/node/tree/be525d7d04ec839fd2bb47affef00bb3c1302aca/src#context
- https://github.com/nodejs/node/tree/be525d7d04ec839fd2bb47affef00bb3c1302aca/src#environment
